Mathlib/Algebra/Polynomial/CoeffList.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2023 Alex Meiburg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Meiburg
+-/
+import Mathlib.Algebra.Polynomial.Degree.Definitions
+import Mathlib.Algebra.Polynomial.EraseLead
+import Mathlib.Data.List.Range
+
+/-!
+# A list of coefficients of a polynomial
+
+## Definition
+
+* `coeffList f`: a `List` of the coefficients, from leading term down to constant term.
+* `coeffList 0` is defined to be `[]`.
+
+This is useful for talking about polynomials in terms of list operations, and applying list theorems
+to them, instead of working with `Finsupp` structures.
+
+The most significant theorem here is `coeffList_eraseLead`, which says that `coeffList P` can be
+written as `(leadingCoeff P) :: (List.replicate k 0) ++ (coeffList P.eraseLead)`. That is, the list
+of coefficients starts with the leading coefficient, followed by some number of zeros, and then the
+coefficients of `P.eraseLead`.
+-/
+open Polynomial
+
+namespace Polynomial
+
+variable {α : Type*}
+
+section Semiring
+
+variable [Semiring α]
+
+/-- A list of coefficients starting from the leading term down to the constant term. Defining
+this with P.support keeps it computable, even if the base ring doesn't have DecidableEq.
+-/
+def coeffList (P : α[X]) : List α :=
+  if P.support = ∅ then [] else (List.range (P.natDegree+1)).reverse.map P.coeff
+
+variable {P : α[X]}
+
+theorem coeffList_of_ne_zero (h : P ≠ 0) :
+    P.coeffList = (List.range (P.natDegree+1)).reverse.map P.coeff := by
+  rw [coeffList, if_neg (mt support_eq_empty.mp h)]
+
+variable (α) in
+@[simp]
+theorem coeffList_zero : (0 : α[X]).coeffList = [] := by
+  simp [coeffList]
+
+/-- Only the zero polynomial gives nil coeffList -/
+@[simp]
+theorem coeffList_eq_nil_iff (P : α[X]) : P.coeffList = [] ↔ P = 0 :=
+  ⟨by simp_all [coeffList], (· ▸ coeffList_zero α)⟩
```

Comment:
and this can be proved by
```suggestion
theorem coeffList_eq_nil_iff (P : α[X]) : P.coeffList = [] ↔ P = 0 := by simp [coeffList]
```
or something similar